StateView & Flash Accounting guides added (#848)
Uniswap/docs: docs/contracts/v4/guides/13-flash-accounting.mdx

# Understanding the Basics

Before diving into implementation patterns, let’s look at the key concepts you’ll need to work with flash accounting. Each example includes common scenarios you’ll encounter when building on v4.

## Working with Deltas

Every operation in v4 that involves tokens creates deltas. These deltas track what the executor owes to the PoolManager and vice versa:

```solidity
// Example: Executing a swap
// Note: This assumes the PoolManager has been unlocked
function executeSwap(PoolKey calldata key) external {
    // A swap returns a BalanceDelta
    BalanceDelta delta = poolManager.swap(
        key,
        IPoolManager.SwapParams({
            zeroForOne: true,
            amountSpecified: -1e18,  // Negative means spending/providing 1 ETH
            sqrtPriceLimitX96: MAX_SQRT_RATIO - 1  // Max price willing to accept
        }),
        ""
    );

    // Delta shows:
    // delta.amount0() = -1e18    (executor owes 1 ETH)
    // delta.amount1() = +2000e6  (executor receives 2000 USDC)
}
```

> When a swap is executed, the PoolManager returns a `BalanceDelta` that shows your token obligations. In this example, the negative delta (-1e18) means the executor owes 1 ETH to the PoolManager, while the positive delta (+2000e6) means the executor is entitled to receive 2000 USDC. These deltas must be resolved before the transaction completes.

Reviewer: Probably could link to the tech reference for `BalanceDelta`

```suggestion
When a swap is executed, the PoolManager returns a [`BalanceDelta`](https://docs.uniswap.org/contracts/v4/reference/core/types/balancedelta) that shows your token obligations. In this example, the negative delta (-1e18) means the executor owes 1 ETH to the PoolManager, while the positive delta (+2000e6) means the executor is entitled to receive 2000 USDC. These deltas must be resolved before the transaction completes.
```

Author: done

> Note how negative values in v4 consistently represent tokens going to the PoolManager - both in `amountSpecified` for the input amount and in the returned delta for tokens owed.

Reviewer: Probably could make this note italic?

```suggestion
_Note how negative values in v4 consistently represent tokens going to the PoolManager - both in `amountSpecified` for the input amount and in the returned delta for tokens owed._
```

## Reading Delta States

A common pattern is checking current deltas before executing operations. The `TransientStateLibrary` helps you track these balances:

```solidity
import {TransientStateLibrary} from "@uniswap/v4-core/contracts/libraries/TransientStateLibrary.sol";

contract DeltaReader {
    using TransientStateLibrary for IPoolManager;

    function checkDeltaBeforeOperation(
        Currency currency,
        address user
    ) external view returns (int256) {
        // Important: This shows the current delta for this token/user pair
        return poolManager.getCurrentDelta(currency, user);
        // Negative: User owes tokens
        // Positive: User can claim tokens
        // Zero: No outstanding obligations
    }
}
```

> The `TransientStateLibrary` provides utilities to check the current state of deltas at any point in your transaction. The `getCurrentDelta` function returns an int256 where negative values indicate the user owes tokens to the PoolManager, positive values mean the user can claim tokens, and zero means there are no outstanding obligations for this token/user pair.

Reviewer: ```suggestion
The `TransientStateLibrary` provides utilities to check the current state of deltas at any point in your transaction. The `getCurrentDelta` function returns an int256 where negative values indicate the user owes tokens to the PoolManager, positive values mean the user can claim tokens from the PoolManager, and zero means there are no outstanding obligations for this token/user pair.
```

    // For positive deltas (receiving tokens):
    poolManager.take(currency, address(this), amount);
}
```

When resolving negative deltas with ERC-20 tokens, you need to:

1. Sync the currency balance with `sync()`
2. Transfer the tokens to the PoolManager
3. Complete the settlement with `settle()`

For positive deltas, simply use `take` to receive tokens from the PoolManager.

**2. Using ERC-6909 Functions**

```solidity
function resolveWithERC6909(
    Currency currency,
    uint256 amount
) external {
    // For negative deltas (you owe tokens):
    poolManager.burn(currency, address(this), amount);

    // For positive deltas (receiving tokens):
    poolManager.mint(currency, address(this), amount);
}
```

ERC-6909 operations map to their ERC-20 equivalents in v4:

- Use `burn` when you would use `settle` (for negative deltas)
- Use `mint` when you would use `take` (for positive deltas)

Notice how this pattern requires no additional sync operations or separate token transfers.

> **Important**: *Every delta must be resolved before the transaction ends, or the entire transaction will revert. Use* `TransientStateLibrary` *to verify your balances are properly settled.*
> 

> /callout Deltas are not associated to a "token type" therefore developers can mix and match the ERC-20 functions and the ERC-6909 functions

Reviewer: Is it better to rephrase it like this?

```suggestion
> _Delta is a net balance resulting from token movements thus not bound to a certain token type i.e. can be resolved via mix-and-match with ERC-20 functions and ERC-6909 functions._
```

# Working with Flash Accounting

To interact with the PoolManager, we first need to create the functions our users will call. Then we'll implement the unlock callback pattern required to execute these operations.

## Using the Lock/Unlock Pattern

Let’s start by creating our external function:

```solidity

> function executeSwap(

Reviewer 2: Extend this code to include `unlockCallback` that the user needs to implement to make it more obvious.

Reviewer 3: agreed, we can stub it like:

```solidity
function unlockCallback(bytes calldata data) external returns (bytes memory) {
    // to be implemented later
}
```

Author: done

    PoolKey calldata key,
    uint256 amount
) external returns (int256, int256) {
    // Encode operation parameters
    bytes memory data = abi.encode(key, amount);

    // Call unlock with encoded data
    bytes memory result = poolManager.unlock(data);

    // Optional: Decode any relevant return data
    return (0, 0); // Replace with actual return values if needed
}
```

When you call this function the flow followed is the following:

1. `unlock` is called on the PoolManager
2. PoolManager calls back to your `unlockCallback`
3. Your callback executes the operations
4. All deltas must be resolved before returning

> 5. Context returns to the PoolManager which verifies there are no outstanding deltas, and will relock itself

Reviewer 2: Could you clarify what you mean by context?

Reviewer 3: yeah i added this suggestion -- maybe we should just say `Execution of the logic returns to the PoolManager...`

Author: done

> **Warning***: Always implement proper access control in your unlock callback. Only the PoolManager should be able to call it.*
> 

## Implementing the Unlock Callback

First, let’s set up a contract with the proper unlock callback implementation:

```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.24;

import {IPoolManager} from "@uniswap/v4-core/contracts/interfaces/IPoolManager.sol";
import {PoolKey} from "@uniswap/v4-core/contracts/types/PoolKey.sol";
import {BalanceDelta} from "@uniswap/v4-core/contracts/types/BalanceDelta.sol";
import {Currency} from "@uniswap/v4-core/contracts/types/Currency.sol";

> contract FlashAccountingExample {
>     IPoolManager public immutable poolManager;
> 
>     constructor(IPoolManager _poolManager) {
>         poolManager = _poolManager;
>     }
> 
>     function unlockCallback(bytes calldata data) external returns (bytes memory) {
>         // Important: Must check caller is PoolManager
>         require(msg.sender == address(poolManager), "Not pool manager");
> 
>         // Decode and call our executeOperations function which 
>         // we'll implement next
>         (bytes memory result) = executeOperations(data);
> 
>         // Important: Must return bytes, even if empty
>         return result;
>     }
> }

Reviewer: Should we also include the `executeSwap()` in this contract for better clarity?

```suggestion
contract FlashAccountingExample {
    IPoolManager public immutable poolManager;

    constructor(IPoolManager _poolManager) {
        poolManager = _poolManager;
    }
    
    function executeSwap(
        PoolKey calldata key,
        uint256 amount
    ) external returns (int256, int256) {
        ...
    }

    function unlockCallback(bytes calldata data) external returns (bytes memory) {
        // Important: Must check caller is PoolManager
        require(msg.sender == address(poolManager), "Not pool manager");

        // Decode and call our executeOperations function which 
        // we'll implement next
        (bytes memory result) = executeOperations(data);

        // Important: Must return bytes, even if empty
        return result;
    }
}
```

```

This base contract sets up the foundation for working with v4’s flash accounting. The `unlockCallback` function is required for any operations that access pool liquidity - when your contract calls `poolManager.unlock()`, the PoolManager calls back to this function to execute your operations.

The callback must verify it's being called by the PoolManager and return a bytes value (even if empty) to prevent transaction failures. Any actual pool operations (like swaps or liquidity changes) will be handled through the `executeOperations` function.

> **Critical Note***: The most common mistake developers make is not returning a bytes value from unlockCallback. This will cause your transaction to revert. Always return a bytes value, even if it’s empty.*
> 

Let’s add functionality to execute operations:

```solidity
function executeOperations(
    bytes calldata data
) internal returns (bytes memory) {
    // Decode operation parameters
    (PoolKey memory key, uint256 amount) = abi.decode(
        data,
        (PoolKey, uint256)
    );

>     // Execute operation (e.g., swap)

Reviewer: ```suggestion
    // Execute operation (e.g. swap)
```
